src/apps/recentlyUsed.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * Gets the list of apps that have been used recently by the authenticated user.
+ *
+ * @module apps/recentlyUsed
+ */
+
+import { validateInterval, validateLimit } from "../util";
+import * as config from "../configuration";
+import constants from "../constants";
+import logger from "../logging";
+
+// All apps returned by this query are DE apps, so the system ID can be constant.
+const recentlyUsedAppsQuery = `
+    SELECT DISTINCT
+        a.id,
+        'de' AS system_id,
+        a.name,
+        a.description,
+        a.wiki_url,
+        a.integration_date,
+        a.edited_date,
+        u.username,
+        EXISTS (
+            SELECT * FROM users authenticated_user
+            JOIN workspace w ON authenticated_user.id = w.user_id
+            JOIN app_category_group acg ON w.root_category_id = acg.parent_category_id
+            JOIN app_category_app aca ON acg.child_category_id = aca.app_category_id
+            WHERE authenticated_user.username = $1
+            AND acg.child_index = $2
+            AND aca.app_id = a.id
+        ) AS is_favorite,
+        TRUE AS is_public,
+        max(j.start_date) AS most_recent_start_date
+    FROM jobs j
+    JOIN users ju ON j.user_id = ju.id
+    JOIN apps a ON CAST(a.id AS text) = j.app_id
+    LEFT JOIN integration_data d ON a.integration_data_id = d.id
+    LEFT JOIN users u ON d.user_id = u.id
+    WHERE ju.username = $1
+    AND NOT a.deleted
+    AND NOT a.disabled
```

Comment:
Poking at a [query plan](https://explain.depesz.com/s/duPn) I wonder if we could benefit from having a partial index on the apps table `WHERE NOT a.deleted AND NOT a.disabled`. I suspect some other stuff would use it as well, and it'd let this use an index scan rather than sequential. At 4.4 milliseconds it's not exactly _slow_, but is the step of the query plan that's taking the largest chunk of time of anything, and it looks like a little less than half of the apps in the prod database are excluded by one or both of these (4307 excluded, 4817 not).

Not required by any means, of course. I might easily be wrong about when it'd use that index, too.

Reply:
I tried adding the index, and Postgres didn't use it. 😞

Reply:
It may have concluded the sequential scan was still faster. It's still grabbing more than half the table, so I guess that makes a certain amount of sense. We can try again later. It's 4 milliseconds, after all :laughing: